oda_api/token.py

```diff
@@ -62,57 +68,154 @@ def decode_oda_token(token: str, secret_key=None, allow_invalid=False) -> dict:
+def rewrite_token(new_token,
+                  token_write_methods: Union[Tuple[TokenLocation], TokenLocation] = None,
+                  force_rewrite=False
+                  ):
+    current_token, discover_method = discover_token_and_method(allow_invalid=True)
+    if current_token is not None:
+        current_decoded_token = decode_oda_token(current_token, allow_invalid=True)
+        current_decoded_token_expires_in_s = current_decoded_token['exp'] - time.time()
+
+        new_decoded_token = decode_oda_token(new_token, allow_invalid=True)
+        new_decoded_token_expires_in_s = new_decoded_token['exp'] - time.time()
+
+        if new_decoded_token_expires_in_s < current_decoded_token_expires_in_s:
+            warning_msg = "The new token will expire before the current one"
+            if force_rewrite:
+                warning_msg += ", but it will be used"
+                logger.warning(warning_msg)
+            else:
+                raise RuntimeError("Expiration time of the refreshed token is lower than "
+                                   "the currently available one")
+
+        current_decoded_token_roles = get_token_roles(current_decoded_token)
+        new_decoded_token_roles = get_token_roles(new_decoded_token)
+        new_roles_difference = set(new_decoded_token_roles) - set(current_decoded_token_roles)
+        current_roles_difference = set(current_decoded_token_roles) - set(new_decoded_token_roles)
+        if new_roles_difference != set() and current_roles_difference == set():
+            logger.warning("The new token has more roles than the current one:\n"
+                           f"roles current token: {current_decoded_token_roles}\n"
+                           f"roles new token: {new_decoded_token_roles}")
+        elif len(new_decoded_token_roles) < len(current_decoded_token_roles):
+            warning_msg = "The new token has less roles than the current one:\n" \
+                          f"roles current token: {current_decoded_token_roles}\n" \
+                          f"roles new token: {new_decoded_token_roles}\n"
+            if force_rewrite:
+                warning_msg += ", but it will be used.\n"
+                logger.warning(warning_msg)
+            else:
+                logger.warning(warning_msg)
+                raise RuntimeError("The roles of the new token are less than those of the current one")
+
+    if token_write_methods is not None:
+        if current_token is not None:
+            with open("old-oda-token_" + str(time.time()), 'w') as ft:
+                ft.write(current_token)
+
+        if isinstance(token_write_methods, TokenLocation):
+            token_write_methods = token_write_methods,
+
+        if discover_method is not None:
+            if discover_method == TokenLocation.ODA_ENV_VAR:
+                environ.pop('ODA_TOKEN', None)
+            elif discover_method == TokenLocation.FILE_CUR_DIR:
+                if path.exists(path.join(getcwd(), ".oda-token")):
```

Comment:
how can it not exist when it was discovered there?
If it does not exist - we should raise RuntimeError because it means there is some internal bug